### DISCUSSION TITLE
ToC - Table of Contents in PDF file

### DISCUSSION THREAD

**Discussion by damian-krawczyk** [marp-team/marp#199] [Ideas] [answered]
Hi,
Does it possible to generate pdf with TOC? Can you add TOC to pdf file generated using Marp for VS Code?

Regards,
Damian Krawczyk

**Comment by yhatt** [comment]
No, it cannot because ToC is not a part of CommonMark. To keep sustainability of Marp ecosystem, Marp team does not have any plans to add the unique features that is far from a common Markdown syntax.

### Extend Marp syntax by using a plugin

[Marp CLI can extend syntax by using third-party plugins.](https://github.com/marp-team/marp-cli#engine) You can try to use [a one of compatible markdown-it plugins for ToC](https://www.npmjs.com/search?q=keywords%3Amarkdown-it-plugin%20toc).

See details in https://github.com/marp-team/marp-core/issues/136#issuecomment-564388487.

```javascript
// An example of engine.js for Marp CLI's custom engine
module.exports = ({ marp }) => marp.use(require('markdown-it-table-of-contents'))
```

> Marp for VS Code is designed for beginners and moderate users. To save them from malicouses, [we are not active for adding an ability of Marp plugins support to VS Code extension.](https://github.com/marp-team/marp/discussions/164#discussioncomment-1313187) Currently you should use Marp CLI whenever want to extend Marp syntax via plugin.

### Markdown All in One VS Code extension

[Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) VS Code extension can generate and add ToC to the current Markdown document.

[<img src="https://github.com/yzhang-gh/vscode-markdown/raw/master/images/toc.png" width="300" />](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one)

### Related

- https://github.com/marp-team/marp-core/issues/136

**Reply by roland-KA** [reply]
I don't quite understand your answer to this request. As to my understanding, there is no need to change or add anything to the markdown syntax, in order to create a toc in PDF. 

It would be sufficient to create a toc in the PDF generated by just using the headlines in the markdown file (so sections using H1, H2 and H3 could be used to generate a toc). Or do I miss something?

I'm talking here about a TOC that most PDF readers display in a side frame, not a TOC that would be part of the document itself.

**Reply by yhatt** [reply]
> I'm talking here about a TOC that most PDF readers display in a side frame, not a TOC that would be part of the document itself.

Thanks to make clear that. I had missed "ToC as PDF metadata" because it should be a part of slide contents to include something into PDF, in current Marp.

Technically we are ready to add [pdf-lib's postprocess](https://github.com/Hopding/pdf-lib/issues/127#issuecomment-502450179) to Marp CLI, to add annotation of the outline metadata into generated PDF by Puppeteer. ([Just like `--pdf-notes` option](https://github.com/marp-team/marp-cli/blob/3db15a2a7df47a6c0d5ea6cc065e3dad81a68304/src/converter.ts#L273-L295)) It requires to know PDF specification about the document outline for making that.

**Reply by roland-KA** [reply]
Thanks for your fast reply!

You write, that you are technically ready to do this. So can we expect in an upcoming release something like a `--pdf-toc` option that would allow us to generate a TOC in PDF?

**Reply by yhatt** [reply]
Yeah, it's well worth to work on that because PDF is a primary format for Marp :)

**Comment by ymollard** [comment]
@damian-krawczyk: in case it helps, I have sketched a [Python script](https://github.com/ymollard/extract-toc-markdown) for that.

(Not sure the Mardown All in One extension works for this: afaik, no anchor is given to titles in Marp, I think we need to link to slides numbers instead)

**Comment by yhatt** [comment]
The request and required tasks have become clear by [community's feedback](https://github.com/orgs/marp-team/discussions/199#discussioncomment-3655355), so I've created https://github.com/marp-team/marp-cli/issues/478. Please see that to check details and track the progress.

**Comment by yhatt** [comment, marked as answer]
**UPDATE:** [Marp CLI v2.2.0](https://github.com/marp-team/marp-cli/releases/tag/v2.2.0) has been released with new `--pdf-outlines` option to make outlines from slide pages and Markdown headings.

```
marp ./markdown.md --pdf --pdf-outlines
```

![](https://user-images.githubusercontent.com/3993388/190893014-f96e53a0-da7b-4460-affc-fb30f9a0fed8.gif)

**Reply by roland-KA** [reply]
Wonderful! Thanks a lot 🤗
